libserial/LibSerial.c

```diff
@@ -36,36 +51,44 @@ EM_ASYNC_JS(void, clear_read_buffer, (int timeoutMs), {
             resolve();
         };
     });
+    window.readAddress = 3;
     window.avrdudeLog = [...window.avrdudeLog, "Read buffer cleared"];
 });
 
 EM_ASYNC_JS(void, read_data, (int timeoutMs, int length), {
-    window.avrDudeWorker.postMessage({ type: 'read', timeout: timeoutMs, requiredBytes: length });
-    const data = await new Promise((resolve, _) => {
-        window.avrDudeWorker.onmessage = (event) => {
-            // check if the response type is an error
-            if (event.data.type === "error") {
-                window.funcs._errorCallback();
-            } else if (event.data.type == "read") {
-                resolve(event.data);
-            }
+    const result = new Uint8Array(length);
+    let read = 0;
+    let end = Date.now() + timeoutMs;
+
+    while (read !== length && end > Date.now()) {
+        if (window.readAddress === window.readAddressBuf[0]) continue;
+
+        let targetAddress = window.readAddressBuf[0];
+        if (targetAddress < window.readAddress) {
+            targetAddress = window.readBuffer.length;
         };
-    });
-    const result = data.result;
-
-    if (result instanceof Uint8Array && result.length > 0) {
-        //console.log("Received: ", result);
-        // convert data into an readable string formated like this 0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15
-        const printResult = Array.from(result).join(",");
-        window["avrdudeLog"] = [...window["avrdudeLog"], "Received: " + printResult];
-        const ptr = window.funcs._malloc(result.length * Uint8Array.BYTES_PER_ELEMENT);
-        window.funcs.HEAPU8.set(result, ptr);
-
-        // Call the C++ function with the pointer and the length of the array
-        window.funcs._dataCallback(ptr, result.length);
-    } else {
+        targetAddress = Math.min(targetAddress, window.readAddress + length - read);
+
+        result.set(window.readBuffer.slice(window.readAddress, targetAddress), read);
+        read += targetAddress - window.readAddress;
+
+        window.readAddress = targetAddress;
+        if (window.readAddress === window.readBuffer.length) {
+            window.readAddress = 3;
+        }
+    }
+
+    if (read === 0) {
```

Comment:
This check can be improved: what if some data was read, and then we run into a timeout?